crates/matrix-sdk/src/latest_events/latest_event.rs

```diff
@@ -150,6 +169,332 @@ impl LatestEventValue {
+                                .push(transaction_id.to_owned(), value.clone());
+
+                            value
+                        } else {
+                            Self::None
+                        }
+                    } else {
+                        Self::None
+                    }
+                }
+
+                LocalEchoContent::React { .. } => Self::None,
+            },
+
+            // A local event has been cancelled before being sent.
+            //
+            // Remove the calculated `LatestEventValue` from the buffer of values, and return the
+            // last `LatestEventValue` or calculate a new one.
+            RoomSendQueueUpdate::CancelledLocalEvent { transaction_id } => {
+                if let Some(position) = buffer_of_values_for_local_events.position(transaction_id) {
+                    buffer_of_values_for_local_events.remove(position);
+                }
+
+                Self::new_local_or_remote(
+                    buffer_of_values_for_local_events,
+                    room_event_cache,
+                    power_levels,
+                )
+                .await
+            }
+
+            // A local event has successfully been sent!
+            //
+            // Unwedge all wedged values after the one matching `transaction_id`. Indeed, if
+            // an event has been sent, it means the send queue is working, so if any value has been
+            // marked as wedged, it must be marked as unwedged. Then, remove the calculated
+            // `LatestEventValue` from the buffer of values. Finally, return the last
+            // `LatestEventValue` or calculate a new one.
+            RoomSendQueueUpdate::SentEvent { transaction_id, .. } => {
+                let position = buffer_of_values_for_local_events.unwedged_after(transaction_id);
+
+                if let Some(position) = position {
+                    buffer_of_values_for_local_events.remove(position);
+                }
+
+                Self::new_local_or_remote(
+                    buffer_of_values_for_local_events,
+                    room_event_cache,
+                    power_levels,
+                )
+                .await
+            }
+
+            // A local event has been replaced by another one.
+            //
+            // Replace the latest event value matching `transaction_id` in the buffer if it exists
+            // (note: it should!), and return the last `LatestEventValue` or calculate a new one.
+            RoomSendQueueUpdate::ReplacedLocalEvent { transaction_id, new_content: content } => {
+                if let Some(position) = buffer_of_values_for_local_events.position(transaction_id) {
+                    if let Ok(content) = content.deserialize() {
+                        if let Some(kind) = find_and_map_any_message_like_event_content(content) {
+                            buffer_of_values_for_local_events.replace_kind(position, kind);
+                        }
```

Comment:
:bug: It's a bit of an edge case, but a suitable event content can become unsuitable after the `ReplacedLocalEvent`. E.g. a `m.room.message` local echo has been replaced with one including a replacement relation; as edits aren't handled (per my previous review comment), there won't be any extracted content here, so we will never replace the local value. In this case, I think we should likely remove it from the buffer altogether  (so `else { buffer….remove(position); }`).

Can you also add a test for this, please?